### DISCUSSION TITLE
DBus exception when trying to connect to a BLE device

### DISCUSSION THREAD

**Discussion by HamsterSqueak** [hbldh/bleak#1504] [Q&A]
I am trying to use the Python bleak libary on a Raspberry Pi 4B running Raspbian Debian GNU/Linux 11 (bullseye), using the internal BLE chip. Here's my code:

```python
import asyncio
from bleak import BleakScanner, BleakClient
import sys


async def main():
    #devices = await BleakScanner.discover(timeout=20)
    d = await BleakScanner.find_device_by_address("XX:0D:XX:F1:XX:FC", timeout=20)            
    try:
        if d.address == "XX:0D:XX:F1:XX:FC":
            async with BleakClient(d, timeout=10) as client:
                res = await client.pair(protection_level=None)
                print(res)

                x = client.is_connected
                print("Connected: {0}".format(x))

                thisCharacteristic = svc.get_characteristic("XXXXXXXX-1234-XXXX-1234-56789XXXXXX")
                # frame = bytearray()
                # frame.append(0x04)


                value = bytes(await client.read_gatt_char(thisCharacteristic))

                frame = bytearray()
                frame.append(value[0] + 1)

                await client.write_gatt_char(thisCharacteristic, frame)

                for service in client.services:
                    print("[Service] {0}: {1}".format(service.uuid, service.description))
                    for char in service.characteristics:
                        if "read" in char.properties:
                            try:
                                value = bytes(await client.read_gatt_char(char.uuid))
                            except Exception as e:
                                value = str(e).encode()
                        else:
                            value = None
                        print(
                            "\t[Characteristic] {0}: ({1}) | Name: {2}, Value: {3} ".format(
                                char.uuid, ",".join(char.properties), char.description, value))
                await client.disconnect()
                await client.unpair()

    except:
        e = sys.exc_info()[0]
        print(e)

        
if __name__ == "__main__":
    asyncio.run(main())
```
This code runs well sometimes, verifying that I can write a value to a characteristic and read values of characterisitcs. Sometimes it has an exception while trying to connect (`async with BleakClient(d, timeout=10) as client:`). The exception is `<class 'bleak.exc.BleakDBusError'>`. After this exception occurs, I can sometimes execute the code successfully (no exception), but other times it has the exception every execution until I reboot the Raspberry Pi.

As this communication process is to occur in a manufacturing environment, I am concerned about this solution.

Can someone help me understand how to get the bleak code to work reliably (no DBus exception)?

**Comment by dlech** [comment]
What is the specific D-Bus error? What is the full stack trace? Can you share debug logs? Can you share packet capture logs?

**Reply by HamsterSqueak** [reply]
I don't know any more than I sent. I included all text that is "printed" from the exception. What do I need to do to get packet capture logs?

**Reply by HamsterSqueak** [reply]
I'll try to include the full stack trace.

**Reply by dlech** [reply]
https://bleak.readthedocs.io/en/latest/troubleshooting.html

**Reply by HamsterSqueak** [reply]
Many thanks. Do I need to sniff using a separate BLE adapter from the one that is used for BLE GATT communication? If so, is it appropriate to run the sniffer on Windows 10 Pro and the BLE GATT communication on a Raspberry Pi 4B? Also, I already have an nRF 52840 dongle, an nRF 52840 DK, and a BlueGiga BLED112 USB to BLE adapter. I didn't seem to be able to use them from the Raspberry Pi for BLE GATT communication. Can one of those be used for sniffing? (I realize that this goes a bit beyond the original request for help.)

**Reply by dlech** [reply]
No sniffer needed. You can just capture the HCI packets sent to/from the adapter on the RPi, e.g. using `btmon`.

But if you are interested in a sniffer anyway, nRF has some firmware that can turn the 5280 into a sniffer. https://www.nordicsemi.com/Products/Development-tools/nrf-sniffer-for-bluetooth-le
